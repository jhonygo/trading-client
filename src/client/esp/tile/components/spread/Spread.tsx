import React from 'react';
import * as style from './style.css';

export namespace Spread {
  export interface Props {
    bid: number;
    ask: number;
  }

  export interface State {
    spread: string;
    direction: string;
  }
}

export class Spread extends React.PureComponent<Spread.Props, Spread.State> {
  constructor(props: Spread.Props, context: any) {
    super(props, context);
    this.state = { spread: '0', direction: 'neutral' };
  }

  componentDidUpdate(prevProps) {
    if (this.props.ask !== prevProps.ask || this.props.bid !== prevProps.bid) {
      const spread = ((this.props.ask - this.props.bid) * 1000).toFixed(2);
      const direction = this.props.ask > prevProps.ask ? 'up' : 'down';
      this.setState({ spread: spread, direction: direction });
    }
  }

  render() {
    const { direction, spread } = this.state;

    const visible = {
      visibility: 'visible'
    } as React.CSSProperties;

    const hidden = {
      visibility: 'hidden'
    } as React.CSSProperties;

    const upStyle =
      direction === 'up' ? visible : (hidden as React.CSSProperties);

    const downStyle =
      direction === 'down' ? visible : (hidden as React.CSSProperties);
    return (
      <div className={style.spread}>
        <div className={style.spreadUp} style={upStyle} />
        <div className={style.spreadValue}>
          <small>{spread}</small>
        </div>
        <div className={style.spreadDown} style={downStyle} />
      </div>
    );
  }
}
