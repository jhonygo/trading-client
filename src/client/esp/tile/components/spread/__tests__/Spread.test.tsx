import React from 'react';
import { shallow } from 'enzyme';
import { Spread } from '../Spread';

describe('Spread', () => {
  it('should update spread value', () => {
    // Given
    const spread = shallow(<Spread bid={0} ask={0} />);

    // When
    spread.setProps({ bid: 1.43841, ask: 1.43848 });

    // Then
    expect(spread.state().spread).toEqual('0.07');
  });

  it('should display spread up direction', () => {
    // Given
    const spread = shallow(<Spread bid={1.43841} ask={1.43847} />);

    // When
    spread.setProps({ bid: 1.43841, ask: 1.43848 });

    // Then
    expect(spread.state().direction).toEqual('up');
    expect(spread).toMatchSnapshot();
  });

  it('should display spread down direction', () => {
    // Given
    const spread = shallow(<Spread bid={1.43841} ask={1.43847} />);

    // When
    spread.setProps({ bid: 1.43841, ask: 1.43846 });

    // Then
    expect(spread.state().direction).toEqual('down');
    expect(spread).toMatchSnapshot();
  });
});
