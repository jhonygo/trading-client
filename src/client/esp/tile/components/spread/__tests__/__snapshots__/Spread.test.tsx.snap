// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spread should display spread down direction 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Spread
    ask={1.43846}
    bid={1.43841}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div>
        <div
          style={
            Object {
              "visibility": "hidden",
            }
          }
        />
        <div>
          0.05
        </div>
        <div
          style={
            Object {
              "visibility": "visible",
            }
          }
        />
      </div>,
      "className": "col-md-4",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            style={
              Object {
                "visibility": "hidden",
              }
            }
          />,
          <div>
            0.05
          </div>,
          <div
            style={
              Object {
                "visibility": "visible",
              }
            }
          />,
        ],
        "className": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": undefined,
            "style": Object {
              "visibility": "hidden",
            },
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "0.05",
            "className": undefined,
          },
          "ref": null,
          "rendered": "0.05",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": undefined,
            "style": Object {
              "visibility": "visible",
            },
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div>
          <div
            style={
              Object {
                "visibility": "hidden",
              }
            }
          />
          <div>
            0.05
          </div>
          <div
            style={
              Object {
                "visibility": "visible",
              }
            }
          />
        </div>,
        "className": "col-md-4",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              style={
                Object {
                  "visibility": "hidden",
                }
              }
            />,
            <div>
              0.05
            </div>,
            <div
              style={
                Object {
                  "visibility": "visible",
                }
              }
            />,
          ],
          "className": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": undefined,
              "style": Object {
                "visibility": "hidden",
              },
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "0.05",
              "className": undefined,
            },
            "ref": null,
            "rendered": "0.05",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": undefined,
              "style": Object {
                "visibility": "visible",
              },
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Spread should display spread up direction 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Spread
    ask={1.43848}
    bid={1.43841}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div>
        <div
          style={
            Object {
              "visibility": "visible",
            }
          }
        />
        <div>
          0.07
        </div>
        <div
          style={
            Object {
              "visibility": "hidden",
            }
          }
        />
      </div>,
      "className": "col-md-4",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            style={
              Object {
                "visibility": "visible",
              }
            }
          />,
          <div>
            0.07
          </div>,
          <div
            style={
              Object {
                "visibility": "hidden",
              }
            }
          />,
        ],
        "className": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": undefined,
            "style": Object {
              "visibility": "visible",
            },
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "0.07",
            "className": undefined,
          },
          "ref": null,
          "rendered": "0.07",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": undefined,
            "style": Object {
              "visibility": "hidden",
            },
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div>
          <div
            style={
              Object {
                "visibility": "visible",
              }
            }
          />
          <div>
            0.07
          </div>
          <div
            style={
              Object {
                "visibility": "hidden",
              }
            }
          />
        </div>,
        "className": "col-md-4",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              style={
                Object {
                  "visibility": "visible",
                }
              }
            />,
            <div>
              0.07
            </div>,
            <div
              style={
                Object {
                  "visibility": "hidden",
                }
              }
            />,
          ],
          "className": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": undefined,
              "style": Object {
                "visibility": "visible",
              },
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "0.07",
              "className": undefined,
            },
            "ref": null,
            "rendered": "0.07",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": undefined,
              "style": Object {
                "visibility": "hidden",
              },
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
